src/main/java/com/zufar/onlinestore/user/api/ConfirmUserEmailOperationPerformer.java

```diff
@@ -0,0 +1,75 @@
+package com.zufar.onlinestore.user.api;
+
+import com.zufar.onlinestore.user.exception.UserNotFoundException;
+import com.zufar.onlinestore.user.repository.UserRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.lang.NonNull;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.UUID;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ConfirmUserEmailOperationPerformer {
+
+    private final UserRepository userCrudRepository;
+
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.READ_COMMITTED)
+    public String generateUserEmailConfirmationToken(final UUID userId) {
+        var userEntity = userCrudRepository.findById(userId)
+                .orElseThrow(() -> {
+                    log.error("Failed to get the user with the id = {}.", userId);
+                    return new UserNotFoundException(userId);
+                });
+        final var token = generateConfirmationToken(userEntity.getId());
+        userEntity.setConfirmationToken(token);
+        userCrudRepository.save(userEntity);
+        return token;
+    }
+
+    public void confirmUserEmail(final String token) {
+        var userEntity = userCrudRepository.findByConfirmationToken(token)
+                .orElseThrow(() -> {
+                    log.error("Failed to get the user with the token = {}.", token);
+                    return new UserNotFoundException(token);
+                });
+        userEntity.setEmailConfirmed(true);
+        userEntity.setConfirmationToken(null);
+        userCrudRepository.save(userEntity);
+    }
+
+    /**
+     * Generates a user-unique confirmation token.
+     * The token is a string of characters and digits based on user's UUID.
+     *
+     * @param userUuid the user's uuid.
+     * @return the confirmation token.
+     */
+    private @NonNull String generateConfirmationToken(@NonNull UUID userUuid) {
```

Comment:
1. Firstly,

I ran this token generation method 2 times for 10,000,000 unique UUIDs
The first time this code generated 9,952,520 unique tokens for 10,000,000 unique UUIDs.
The second time, this code generated 9,952,345 unique tokens for 10,000,000 unique UUIDs.
Thus, about 50,000 tokens are not unique and are duplicated. The collision percentage is 0.5%, which is very high, meaning every two hundredth user has the same verification token.

Is this level of collision acceptable?

2. Secondly,

The token generation method is based on the user ID and is a constant for the user on any run.

This means that if someone finds out the user’s UUD identifier and recognizes the algorithm (for example, an unscrupulous developer), he can easily generate a token for any operation on behalf of another user. Secret tokens should not be based on known data and should not be subject to brute force.

3. Thirdly,

The generation algorithm contains "magic numbers".
What are 26, 65, 97 ?

4. Tokens have different lengths, so it will not be possible to limit the size of a token in the database, but simply specify a VARCHAR without a size;

Reply:
```
class SecureRandomTokenGenerator {

    private static final String NUMERIC_BASE = "0123456789";
    private static final String LATIN_BASE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    private static final String DEFAULT_BASE = NUMERIC_BASE + LATIN_BASE;

    private final SecureRandom random = new SecureRandom();

    public String nextToken(int tokenLength) {
        StringBuilder token = new StringBuilder(tokenLength);
        for (int i = 0; i < tokenLength; i++) {
            int randomSymbolIndex = random.nextInt(DEFAULT_BASE.length());
            token.append(DEFAULT_BASE.charAt(randomSymbolIndex));
        }
        return token.toString();
    }

    public String nextToken(String base, int tokenLength) {
        StringBuilder token = new StringBuilder(tokenLength);
        for (int i = 0; i < tokenLength; i++) {
            int randomSymbolIndex = random.nextInt(base.length());
            token.append(base.charAt(randomSymbolIndex));
        }
        return token.toString();
    }

    /**
     * Example,
     * Base is "0123456789"
     * Replace hook is '#'
     * Pattern is "###-###"
     * Instead of '#', generated symbols are substituted and for example, random result will be 143-376
     * Example,
     * Base is "ABCDEFG"
     * Replace hook is '?'
     * Pattern is "??-??-??"
     * Instead of '?', generated symbols are substituted and for example, random result will be FD-GA-AB
     */
    public String nextToken(String base, String pattern, char replaceHook) {
        PatternBuilder token = new PatternBuilder(pattern, replaceHook);
        while (token.isReplaceable()) {
            int randomSymbolIndex = random.nextInt(base.length());
            token.replace(base.charAt(randomSymbolIndex));
        }
        return token.toString();
    }
}

class PatternBuilder {

    private final List<Integer> replacingIndexes;
    private final StringBuilder replacedPattern;
    private int replacedSybmolsCount;

    public PatternBuilder(String pattern, Character replaceHook) {
        replacedPattern = new StringBuilder(pattern);
        replacingIndexes = new ArrayList<>();
        replacedSybmolsCount = 0;
        parsePattern(pattern, replaceHook);
    }

    private void parsePattern(String pattern, Character replaceHook) {
        int index = pattern.indexOf(replaceHook);
        while (index >= 0) {
            replacingIndexes.add(index);
            index = pattern.indexOf(replaceHook, index + 1);
        }
    }

    public void replace(char replacingSymbol) {
        Integer replacingIndex = replacingIndexes.get(replacedSybmolsCount);
        replacedPattern.setCharAt(replacingIndex, replacingSymbol);
        replacedSybmolsCount++;
    }

    public boolean isReplaceable() {
        return replacedSybmolsCount < replacingIndexes.size();
    }

    @Override
    public String toString() {
        return replacedPattern.toString();
    }
}
```

Reply:
1. For example, `nextToken(16)` will generate random 16 symbols length tokens and there are no even one collision on 100_000_000 calls.
It is usefull for email link confirmation, but if you want to confirm using short codes like "123-456" there is second method `nextToken(String base, String pattern, char replaceHook)`

2. For example,` nextToken("0123456789", "###-###", '#')`  will generate random 6 symbols length token according to pattern "###-###", for example, "123-456"

Reply:
Thanks for the review, I'll use the snippet above for token generation then